src/merkle_tree.rs

```diff
@@ -0,0 +1,516 @@
+        }
+
+        Some(
+            parent_level
+                .chunks_exact(2)
+                .map(Self::merkle_parent)
+                .collect(),
+        )
+    }
+
+    /// Computes the Merkle root hash for the provided leaf hashes.
+    pub fn root(&self) -> Hash {
+        self.levels.last().unwrap().first().unwrap().clone()
+    }
+
+    /// Hash the provided bytes using SHA-256.
+    /// Returns the hash as a 32 bytes array.
+    ///
+    /// # Examples
+    /// ```
+    /// use merkle_tree::MerkleTree;
+    ///
+    /// let input = "In a hole in the ground there lived a hobbit.";
+    /// let hash = MerkleTree::hash(input.as_bytes());
+    /// ```
+    pub fn hash(bytes: &[u8]) -> Hash {
+        hmac_sha256::Hash::hash(bytes)
+    }
+
+    // Returns tuple (level, index, hash).
+    fn get_parent(&self, level: usize, index: usize) -> Option<TreePosition> {
+        let parent_index = index / 2;
+        let parent_level = level + 1;
+        let parent = self.levels.get(parent_level)?.get(parent_index)?.clone();
+
+        Some(TreePosition {
+            level: parent_level,
+            index: parent_index,
+            hash: parent,
+        })
+    }
+
+    fn get_sibling(&self, level: usize, index: usize) -> Option<TreePosition> {
+        let sibling_index = if index % 2 == 1 { index - 1 } else { index + 1 };
+
+        let sibling = self.levels.get(level)?.get(sibling_index)?.clone();
+
+        Some(TreePosition {
+            level: level,
+            index: sibling_index,
+            hash: sibling,
+        })
+    }
+
+    pub fn proof_of_inclusion(&self, hash: &Hash) -> Option<Vec<Hash>> {
+        let index = self.levels.get(0)?.iter().position(|&h| h == *hash)?;
+
+        let mut current = TreePosition {
+            level: 0,
+            index,
+            hash: hash.clone(),
+        };
+
+        let mut proof: Vec<Hash> = Vec::new();
+
+        while let Some(parent) = self.get_parent(current.level, current.index) {
+            let sibling = self.get_sibling(current.level, current.index);
+            proof.push(sibling.or(Some(current)).unwrap().hash);
+            current = parent;
+        }
+
+        Some(proof)
+    }
+
+    pub fn validate_proof(&self, hash: &Hash, proof: &[Hash]) -> bool {
+        let validation_root = proof.iter().fold(hash.clone(), |hash, sibling| {
+            Self::merkle_parent(&[hash, *sibling])
+        });
+
+        validation_root == self.root()
+    }
+
+    pub fn contains_hash(&self, hash: &Hash) -> bool {
+        self.levels[0].iter().any(|h| h == hash)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn test_hash_should_return_sha256_digest() {
+        let input = "In a hole in the ground there lived a hobbit.";
+        let hash = MerkleTree::hash(input.as_bytes());
+
+        assert_eq!(
+            hash.to_vec(),
+            hex::decode("38a76005681abd4a4f50a364d472016436f17e79778577ee5825580f06997202")
```

Comment:
Here in the tests could be clearer if the different hashes are in variables

Reply:
Got it! Will do.